src/redis/fixed_cache_impl.go

```diff
@@ -69,11 +79,33 @@ func (this *fixedRateLimitCacheImpl) DoLimit(
 				logger.Debugf("cache key is over the limit: %s", cacheKey.Key)
 			}
 			isOverLimitWithLocalCache[i] = true
-			hitAddendForRedis = 0
+			hitsAddendForRedis = 0
 			overlimitIndex = i
+			isCacheKeyOverlimit = true
 			continue
+		} else {
+			// Only if none of the cache key is over limit, call redis to check whether it is getting overlimited.
+			if !isCacheKeyOverlimit {
+				if this.perSecondClient != nil && cacheKey.PerSecond {
```

Comment:
Looks like that it might send multiple requests to redis first for overlimit verification purpose, for every single ratelimit request. Maybe we should balance on the accuracy on this specific use case VS the performance impact. 

Alternatively, we can have a feature flag regarding whether to enable overlimit verification. Besides, when enabling overlimit verification, probably we can use batchGet or pipeline to reduce the perf impact.

Reply:
Yes I will do the suggested approaches to improve the perf.  BTW in previous PR, we have discussed to not to have that feature flag. Anyway when considering the perf impact, I also think its better to have that flag although we considered this as a bug. WDYT @mattklein123 ? Shall we have the feature flag[1] ?

[1] https://github.com/envoyproxy/ratelimit/pull/430#discussion_r1231461436

Reply:
If there are concerns I would just have a config knob and clearly document it. Thank you.

Reply:
@freedomljc , @mattklein123 Added the optimization and perf improvement in 0184b1abc6e72f9d7ea1ee939f0733338c6bbc77 and added the new config and documentation in 42ffea181c2849a7014dd89a680ad07e4dda0ac1